pilot/pkg/networking/plugin/authz/conversion.go

```diff
@@ -0,0 +1,169 @@
+		}
+	case key == attrDestPort:
+		converter = func(v string) (*envoy_rbac.Permission, error) {
+			portValue, err := convertToPort(v)
+			if err != nil {
+				return nil, err
+			}
+			return rbacfilter.PermissionDestinationPort(portValue), nil
+		}
+	case key == pathHeader || key == methodHeader || key == hostHeader:
+		converter = func(v string) (*envoy_rbac.Permission, error) {
+			m := matcher.HeaderMatcher(key, v)
+			return rbacfilter.PermissionHeader(m), nil
+		}
+	case strings.HasPrefix(key, attrRequestHeader):
+		header, err := extractNameInBrackets(strings.TrimPrefix(key, attrRequestHeader))
+		if err != nil {
+			rbacLog.Errorf("ignored invalid %s: %v", attrRequestHeader, err)
+			return nil
+		}
+		converter = func(v string) (*envoy_rbac.Permission, error) {
+			m := matcher.HeaderMatcher(header, v)
+			return rbacfilter.PermissionHeader(m), nil
+		}
+	case key == attrConnSNI:
+		converter = func(v string) (*envoy_rbac.Permission, error) {
+			m := matcher.StringMatcher(v)
+			return rbacfilter.PermissionRequestedServerName(m), nil
+		}
+	case strings.HasPrefix(key, "experimental.envoy.filters.") && isKeyBinary(key):
+		// Split key of format experimental.envoy.filters.a.b[c] to [envoy.filters.a.b, c].
+		parts := strings.SplitN(strings.TrimSuffix(strings.TrimPrefix(key, "experimental."), "]"), "[", 2)
+		converter = func(v string) (*envoy_rbac.Permission, error) {
+			// If value is of format [v], create a list matcher.
+			// Else, if value is of format v, create a string matcher.
+			var m *envoy_matcher.MetadataMatcher
+			if strings.HasPrefix(v, "[") && strings.HasSuffix(v, "]") {
+				m = matcher.MetadataListMatcher(parts[0], parts[1:], strings.Trim(v, "[]"))
+			} else {
+				m = matcher.MetadataStringMatcher(parts[0], parts[1], matcher.StringMatcher(v))
+			}
+			return rbacfilter.PermissionMetadata(m), nil
+		}
+	default:
+		if !found(key, []string{attrDestName, attrDestNamespace, attrDestUser}) &&
+			!strings.HasPrefix(key, attrDestLabel) {
+			// The attribute is neither matched here nor in previous stage, this means it's something we
+			// don't understand, most likely a user typo.
+			rbacLog.Errorf("ignored unsupported constraint: %s", key)
+		}
+		return nil
+	}
+
+	pg := rbacfilter.PermissionGenerator{}
+	for _, v := range values {
+		if rule, err := converter(v); err != nil {
+			rbacLog.Errorf("ignored invalid constraint value: %v", err)
+		} else {
+			pg.Append(rule)
+		}
+	}
+	return pg.OrPermissions()
+}
+
+// principalForKeyValues converts an Istio first class field (e.g. namespaces, groups, ips, as opposed
+// to properties fields) to Envoy RBAC config and returns said field as a Principal or returns nil
+// if the field is empty.
+func principalForKeyValues(key string, values []string, forTCPFilter bool) *envoy_rbac.Principal {
+	pg := rbacfilter.PrincipalGenerator{}
+	for _, value := range values {
+		id := principalForKeyValue(key, value, forTCPFilter)
```

Comment:
nit: maybe there is a better name than `id`? Same with other places that are using `principalForKeyValues`.

Reply:
Renamed to `principal`.